refactor(semantic): add strict mode in scope flags for class definitions

Dunqing opened the pull request:
related: https://github.com/oxc-project/oxc/issues/4142#issuecomment-2219125356

Although we called `enter_node(Class)`, that doesn't mean we're in the `class` scope. It causes our must to visit decorators before `enter_node`.

Let's look at this case. It causes a syntax error if we don't visit decorators before `enter_node`
```js
// This file was procedurally generated from the following sources:
// - src/decorator/decorator-call-expr-identifier-reference-yield.case
// - src/decorator/syntax/valid/cls-expr-decorators-valid-syntax.template
/*---
description: Decorator @ DecoratorCallExpression (Valid syntax for decorator on class expression)
esid: prod-ClassExpression
features: [class, decorators]
flags: [generated, noStrict]
info: |
    ClassExpression[Yield, Await] :
      DecoratorList[?Yield, ?Await]opt class BindingIdentifier[?Yield, ?Await]opt ClassTail[?Yield, ?Await]

    DecoratorList[Yield, Await] :
      DecoratorList[?Yield, ?Await]opt Decorator[?Yield, ?Await]

    Decorator[Yield, Await] :
      @ DecoratorMemberExpression[?Yield, ?Await]
      @ DecoratorParenthesizedExpression[?Yield, ?Await]
      @ DecoratorCallExpression[?Yield, ?Await]

    ...


    DecoratorCallExpression[Yield, Await] :
      DecoratorMemberExpression[?Yield, ?Await] Arguments[?Yield, ?Await]

    DecoratorMemberExpression[Yield, Await] :
      IdentifierReference[?Yield, ?Await]
      DecoratorMemberExpression[?Yield, ?Await] . IdentifierName
      DecoratorMemberExpression[?Yield, ?Await] . PrivateIdentifier

    IdentifierReference[Yield, Await] :
      [~Yield] yield
      ...

---*/
function decorator() {
  return () => {};
}
var yield = decorator;



var C = @yield() class {};

```
Errors:
```shell
  × The keyword 'yield' is reserved
    ╭─[language/statements/class/decorator/syntax/valid/decorator-call-expr-identifier-reference-yield.js:45:2]
 44 │ 
 45 │ @yield() class C {}
    ·  ─────
    ╰────
```


The changed code makes more sense. Only if we call `enter_scope` for class, the flags will contain `StrictMode`. Also, we can get the exact `flags` of the `scope` in the `class` at the transformer

For example:

```
class A {
   B() {
       // Before: flags is `Function`
      //  After: flags is `Function | StrictMode`
   }
}
```

The regression tests will be fixed in follow-up PRs

Dunqing commented:
* **#4156** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4156?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented:
I can't keep up with who should review this anymore, can you request the right person to review?

overlookmotel commented:
I'm afraid I don't have time this week to get into this, but a few thoughts:

1. Can we solve #3862 first? The logic around strict mode in scopes is a bit of a mess, so it's hard to understand whether further changes to it are correct or not.

2. Strict mode also applies to the `extends` clause as well as the class body:

```js
// Sloppy mode
var yield = class {};
class X extends yield {}
// SyntaxError: Unexpected strict mode reserved word
```

3. I don't understand this: *Should* this be a syntax error or not?

```js
// Sloppy mode
var C = @yield() class {};
```

i.e. is the decorator considered part of the class itself, or just sugar for:

```js
var C = yield()(class {});
```

Dunqing commented:
> I'm afraid I don't have time this week to get into this properly, but a few thoughts:
> 
> 1. Can we solve [Inherit `ScopeFlags::StrictMode` from parent scope #3862](https://github.com/oxc-project/oxc/issues/3862) first? The logic around strict mode in scopes is a bit of a mess, so it's hard to understand whether further changes to it are correct or not.

Oh, I just saw this issue and this PR would solve the second point of https://github.com/oxc-project/oxc/issues/3862#issuecomment-2200754940. The first point I can create another PR to fix

> 2. Strict mode also applies to the `extends` clause as well as the class body:
> 
> ```js
> // Sloppy mode
> var yield = class {};
> class X extends yield {}
> // SyntaxError: Unexpected strict mode reserved word
> ```
> 
Yes

> 3. I don't understand this: _Should_ this be a syntax error or not?
> 
> ```js
> // Sloppy mode
> var C = @yield() class {};
> ```
> 

Should not be a syntax error. Because `@yield` is not inside class 

> i.e. is the decorator considered part of the class itself, or just sugar for:
> 
> ```js
> var C = yield()(class {});
> ```

You can think of it as sugar

overlookmotel commented:
> Should not be a syntax error. Because @yield is not inside class

In that case, Babel's decorator transform is incorrect ([Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABAEwKYTgJwIZSwCgEpEBvAKEUU1ShEySMQF4A-UgXwG4z2yA3bJkQBPGKgA2yZinRZcWbv0GIAwtIACoickYRx2AM4GO3IA&debug=false&forceAllTransforms=false&modules=commonjs&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.24.7&externalPlugins=%40babel%2Fplugin-proposal-decorators%407.24.7&assumptions=%7B%7D)). Should we raise an issue on Babel?

overlookmotel commented:
> Looks correct in the latest version [Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABAEwKYTgJwIZSwCgEpEBvAKEUU1ShEySMQF4A-UgXwG4z2yA3bJkQBPGKgA2yZinRZcWbv0GIAwtIACoickYRx2AM4GO3IA&debug=false&forceAllTransforms=false&modules=commonjs&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=true&sourceType=script&lineWrap=true&presets=stage-3&prettier=true&targets=&version=8.0.0-alpha.2&externalPlugins=&assumptions=%7B%7D)

Ah ha yes you're right.

Dunqing commented on tasks/coverage/parser_typescript.snap:
```diff
@@ -13073,22 +13062,6 @@ Expect to Parse: "conformance/salsa/plainJSRedeclare3.ts"
-  × Delete of an unqualified identifier in strict mode.
-   ╭─[conformance/es6/computedProperties/computedPropertyNames3_ES5.ts:5:17]
- 4 │     static [() => { }]() { }
- 5 │     get [delete id]() { }
```
Yes, fixed in https://github.com/oxc-project/oxc/pull/4195

Dunqing committed:
refactor(semantic): add strict mode in scope flags for class definitions (#4156)

related: https://github.com/oxc-project/oxc/issues/4142#issuecomment-2219125356

Although we called `enter_node(Class)`, that doesn't mean we're in the `class` scope. It causes our must to visit decorators before `enter_node`.

Let's look at this case. It causes a syntax error if we don't visit decorators before `enter_node`
```js
// This file was procedurally generated from the following sources:
// - src/decorator/decorator-call-expr-identifier-reference-yield.case
// - src/decorator/syntax/valid/cls-expr-decorators-valid-syntax.template
/*---
description: Decorator @ DecoratorCallExpression (Valid syntax for decorator on class expression)
esid: prod-ClassExpression
features: [class, decorators]
flags: [generated, noStrict]
info: |
    ClassExpression[Yield, Await] :
      DecoratorList[?Yield, ?Await]opt class BindingIdentifier[?Yield, ?Await]opt ClassTail[?Yield, ?Await]

    DecoratorList[Yield, Await] :
      DecoratorList[?Yield, ?Await]opt Decorator[?Yield, ?Await]

    Decorator[Yield, Await] :
      @ DecoratorMemberExpression[?Yield, ?Await]
      @ DecoratorParenthesizedExpression[?Yield, ?Await]
      @ DecoratorCallExpression[?Yield, ?Await]

    ...

    DecoratorCallExpression[Yield, Await] :
      DecoratorMemberExpression[?Yield, ?Await] Arguments[?Yield, ?Await]

    DecoratorMemberExpression[Yield, Await] :
      IdentifierReference[?Yield, ?Await]
      DecoratorMemberExpression[?Yield, ?Await] . IdentifierName
      DecoratorMemberExpression[?Yield, ?Await] . PrivateIdentifier

    IdentifierReference[Yield, Await] :
      [~Yield] yield
      ...

---*/
function decorator() {
  return () => {};
}
var yield = decorator;

var C = @yield() class {};

```
Errors:
```shell
  × The keyword 'yield' is reserved
    ╭─[language/statements/class/decorator/syntax/valid/decorator-call-expr-identifier-reference-yield.js:45:2]
 44 │
 45 │ @yield() class C {}
    ·  ─────
    ╰────
```

The changed code makes more sense. Only if we call `enter_scope` for class, the flags will contain `StrictMode`. Also, we can get the exact `flags` of the `scope` in the `class` at the transformer

For example:

```
class A {
   B() {
       // Before: flags is `Function`
      //  After: flags is `Function | StrictMode`
   }
}
```

The regression tests will be fixed in follow-up PRs
